### PR TITLE
br: recover all tikv data into a consistency state

### LINKED ISSUE #36215: phase 2 restore - recovery all tikv data into a consistency state
after phase 1 restore done, phase 2 restore will do:
1, pull the backup meta data
2, get number of tikv and resolved_ts from meta data
3, stop gc
4, read all region info from tikvs
5, make a recovery plan to tikv
6, resolved kv data by resolved_ts

### PR DESCRIPTION
**Pull request by fengou1**
### What problem does this PR solve?
Issue Number: close #36215

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
https://github.com/fengou1/blr/tree/main/scripts/idc
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/task/restore_data.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.
```
Comment (hawkingrei) [inline br/pkg/task/restore_data.go:1]: 2022

File: br/pkg/restore/data.go [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+	recoveryPlan []*recovpb.RecoverCmdRequest
+}
+
+func NewRecoveryPlan(storeId uint64) RecoveryPlan {
+	var meta = make([]*recovpb.RecoverCmdRequest, 0)
+	return RecoveryPlan{storeId, meta}
+}
+
+type Recovery struct {
+	instances    int
```
Comment (D3Hunter) [inline br/pkg/restore/data.go:84]: prefer tikvCount or storeCount

File: br/pkg/restore/data.go [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+	instanceId   *uint32
```
Comment (D3Hunter) [inline br/pkg/restore/data.go:85]: not used?

File: br/pkg/restore/data.go [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+	regionMetas  []RecoveryMeta
+	recoveryPlan map[uint64][]*recovpb.RecoverCmdRequest
+	resolvedTs   *uint64
```
Comment (D3Hunter) [inline br/pkg/restore/data.go:88]: no need to use pointer

File: br/pkg/restore/data.go [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+func NewRecovery(instances int, tls *tls.Config) Recovery {
+	var instanceId = new(uint32)
+	var regionMetas = make([]RecoveryMeta, instances)
+	var regionRecovers = make(map[uint64][]*recovpb.RecoverCmdRequest, instances)
+	var resolvedTs = new(uint64)
+	return Recovery{instances, instanceId, regionMetas, regionRecovers, resolvedTs, tls}
+}
+
+func (recovery Recovery) newTiKVRecoveryClient(ctx context.Context, tikvAddr string) (recovpb.RecoverDataClient, error) {
```
Comment (D3Hunter) [inline br/pkg/restore/data.go:100]: maybe use `*Recovery` as method receiver

File: br/pkg/restore/data.go [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    time.Duration(10) * time.Second,
+			Timeout: time.Duration(3) * time.Second,
+		}),
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	//defer conn.Close()
```
Comment (D3Hunter) [inline br/pkg/restore/data.go:126]: make sure it's closed

File: br/pkg/task/restore_data.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		return 0, 0, errors.Trace(err)
+	}
+
+	return metaInfo.ClusterInfo.ResolvedTS, metaInfo.TiKVComponent.Replicas, nil
+}
+
+// RunRestore starts a restore task inside the current goroutine.
+func RunRestoreData(c context.Context, g glue.Glue, cmdName string, cfg *RestoreDataConfig) error {
+	var finished bool
+	var totalTiKV int
```
Comment (D3Hunter) [inline br/pkg/task/restore_data.go:99]: duplicate with numOfStores

File: br/pkg/restore/data.go (hunk 4 above)
Comment (fengou1) [reply br/pkg/restore/data.go:87]: removed instanceId, and others for gRPC used.

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**